### ISSUE TITLE
[BUG?] UndeliverableException

### ISSUE THREAD

**Issue by Andresteve07** [patrickvankann/rxmqtt#21]
Hello. I'm facing a situation I can't quite figure how to handle.
In our application we must subscribe to several topics. In the event of network reconnection those subscriptions are disposed by clearing a CompositeDisposable and then fresh new subscriptions to the same topics are created... 
However, long after disposing the first subscriptions when the connection is lost again and the MqttClient can't reconnect then an UndeliverableException is thown that crashes the app.
How should I manage topics re-subuscription?

**Comment by patrickvankann** [comment]
I'm not sure I fully understand the situation with the description above. Is there a gist or repo where I can see this and reproduce the problem?

It would help to know where the UndeliverableException is being thrown. It may be that this exception needs to be handled within the Rx wrapper and converted into an error callback.

I don't understand why this crashes the app though - as a workaround, can you not catch the exception and handle it?

**Comment by Andresteve07** [comment]
Thank you for your response Patrick.
Perhaps I should clarify the test case step by step:
1. WiFi is ON: All subscriptions working!
2. WiFi is OFF
3. MOBILE ON **without** internet access
4. Reconnection routine starts...
given that the MqttAsyncClient is configured to ```setAutomaticReconnect(true)```  and each of the subscriptions try to connect the ObservableMqttClient first, then all are terminated
```onError```  by a paho exception:
```
E/ConnectFactory: Conexión ya en curso  [Connection in progress]
                         Conexión ya en curso (32110)
                             at org.eclipse.paho.client.mqttv3.MqttAsyncClient.connect(MqttAsyncClient.java:734)
                             at net.eusashead.iot.mqtt.paho.ConnectFactory.lambda$create$0$ConnectFactory(ConnectFactory.java:61)
                             at net.eusashead.iot.mqtt.paho.ConnectFactory$$Lambda$0.subscribe(Unknown Source)
                             ...
```
5. All subscriptions are dead now
6. then on automatic reconnection attempt an exception is thrown
```
E/itterMqttActionListener: Se ha perdido la conexión  [Connection is lost]
                          Se ha perdido la conexión (32109) - java.net.SocketException: Connection timed out
                              at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:181)
                              ...
```
7. At this point an UndeliverableException is thrown and the app crashes.
```
E/AndroidRuntime: FATAL EXCEPTION: MQTT Rec: 543874623893
                         Process: com.applicatt.prod.debug, PID: 16376
                         io.reactivex.exceptions.UndeliverableException: Se ha perdido la conexión (32109) - java.net.SocketException: Connection timed out
                            ...
```
I guess the same behaviour should repeat should the phone reconnect to a WiFi AP **without** internet access.
Finally, I copy you a gist with relevant code snippets and complete stacktraces:
https://gist.github.com/Andresteve07/28511712a02584ce837c0a70344f0a75

I hope the test case is clearer now. For more information I am at your disposal.

**Comment by patrickvankann** [comment]
I'm not really sure why Paho throws the exceptions it is throwing - presumably, it is unable to connect to the broker. If your question is why those exceptions are being thrown you'd probably be better served asking on a a Paho forum. 

However, I would expect the Rx wrapper to correctly handle exceptions thrown by Paho and invoke the error handler callbacks without allowing the exception to leak.

From what I can see in the stacktrace, the onFailure handler (in BaseEmitterMqttActionListener) is called which then calls onError() in CompletableEmitterMqttActionListener.  Although the line numbers in the stacktrace don't exactly line up with the current source code in Github - are you using an older version than the current one?

I would have thought this should then call the error handler you have specified in line 21 in your gist. But it looks like the error is somehow propagating to the Android runtime instead, which I presume is causing the Android app to crash. However, I am not an Android expert at all and it's hard to tell from the limited snippets and information here. Could you confirm which version you are using and whether the error handler is being invoked or not (line 21 in the gist). Thanks!

**Comment by patrickvankann** [comment]
Hi Andre,

Sorry that it's been a while but I have just finally had some time to investigate reconnection scenarios a bit further. 

Have you tried using setCleanSession(false) in addition to setAutomaticReconnect(true)?

```
MqttConnectOptions options = new MqttConnectOptions();
options.setAutomaticReconnect(true);
options.setCleanSession(false);
```
This will preserve subscriptions on reconnect assuming you have a persistent store and stable client IDs between reconnects. The behaviour of Paho is described pretty well on this link here:

https://github.com/eclipse/paho.mqtt.java/issues/9 

I was able to simulate a poor connection in an automated test using Toxiproxy to try to replicate what you are describing above so I could reproduce your error.

I found that with the settings above I could:
1. Connect to the broker via the proxy
2. Set up a subscription and publish a message to it
3. Disable the proxy (which would simulate the network being down)
4. Verify that the clients had indeed disconnected after the keep alive timeout
5. Restart the proxy (which would simulate the network being available again)
6. Verify that the clients had reconnected
7. Publish a new message
8. Receive the message on the subscription I had created before the disconnect WITHOUT needing to recreate it

This might help avoid the more complicated approach you have taken to reset subscriptions on reconnect. I didn't receive the error you have described above. 

Hope this helps.
